core/trino-main/src/test/java/io/trino/operator/TestDeduplicationExchangeClientBuffer.java

```diff
@@ -0,0 +1,443 @@
+        }
+
+        try (ExchangeClientBuffer buffer = new DeduplicationExchangeClientBuffer(directExecutor(), ONE_KB, RetryPolicy.QUERY)) {
+            ListenableFuture<Void> blocked = buffer.isBlocked();
+            assertFalse(blocked.isDone());
+
+            TaskId taskId = createTaskId(0, 0);
+            buffer.addTask(taskId);
+            assertFalse(blocked.isDone());
+
+            buffer.taskFinished(taskId);
+            assertFalse(blocked.isDone());
+
+            buffer.noMoreTasks();
+            assertTrue(blocked.isDone());
+        }
+
+        try (ExchangeClientBuffer buffer = new DeduplicationExchangeClientBuffer(directExecutor(), ONE_KB, RetryPolicy.QUERY)) {
+            ListenableFuture<Void> blocked = buffer.isBlocked();
+            assertFalse(blocked.isDone());
+
+            TaskId taskId = createTaskId(0, 0);
+            buffer.addTask(taskId);
+            assertFalse(blocked.isDone());
+
+            buffer.noMoreTasks();
+            assertFalse(blocked.isDone());
+
+            buffer.taskFinished(taskId);
+            assertTrue(blocked.isDone());
+        }
+
+        try (ExchangeClientBuffer buffer = new DeduplicationExchangeClientBuffer(directExecutor(), ONE_KB, RetryPolicy.QUERY)) {
+            ListenableFuture<Void> blocked = buffer.isBlocked();
+            assertFalse(blocked.isDone());
+
+            TaskId taskId = createTaskId(0, 0);
+            buffer.addTask(taskId);
+            assertFalse(blocked.isDone());
+
+            buffer.taskFailed(taskId, new RuntimeException());
+            assertFalse(blocked.isDone());
+
+            buffer.noMoreTasks();
+            assertTrue(blocked.isDone());
+        }
+
+        try (ExchangeClientBuffer buffer = new DeduplicationExchangeClientBuffer(directExecutor(), ONE_KB, RetryPolicy.QUERY)) {
+            ListenableFuture<Void> blocked = buffer.isBlocked();
+            assertFalse(blocked.isDone());
+
+            TaskId taskId = createTaskId(0, 0);
+            buffer.addTask(taskId);
+            assertFalse(blocked.isDone());
+
+            buffer.noMoreTasks();
+            assertFalse(blocked.isDone());
+
+            buffer.taskFailed(taskId, new RuntimeException());
+            assertTrue(blocked.isDone());
+        }
+
+        try (ExchangeClientBuffer buffer = new DeduplicationExchangeClientBuffer(directExecutor(), ONE_KB, RetryPolicy.QUERY)) {
+            ListenableFuture<Void> blocked1 = buffer.isBlocked();
+            ListenableFuture<Void> blocked2 = buffer.isBlocked();
+            assertFalse(blocked1.isDone());
+            assertFalse(blocked2.isDone());
+
+            blocked2.cancel(true);
+
+            assertFalse(blocked1.isDone());
+            assertTrue(blocked2.isDone());
+        }
+    }
+
+    @Test
+    public void testPollPage()
+    {
+        testPollPages(ImmutableListMultimap.of(), ImmutableMap.of(), ImmutableList.of());
+        testPollPages(
+                ImmutableListMultimap.of(
```

Comment:
nit: I would prefre `IM.builder().put(...).put(...).build()`. A bit longer but easier to read IMO.

Reply:
Actually maybe shorter; as you put key and value in same line.